File: ruby/rps.rb

choices = {
  # choice => beats
  "paper"    => "rock",
  "rock"     => "scissors",
  "scissors" => "paper"
}

# play until someone wins
while true

  print "Your move: \n"

  # calculate computer move
  computerMove = choices.keys[rand(choices.length)]

  # get vlid player move
  humanMove = gets.downcase.strip.chomp

  if choices[humanMove] != ""
    if humanMove == computerMove
      # play until someome wins
      print "Tie, Replay!\n"
    elsif choices[humanMove] == computerMove
      print "Human Wins!\n"
      break
    elsif choices[computerMove] == humanMove
      print "Computer Wins!\n"
      break
    end
  end
end

File: node/rps.js
process.stdin.resume();
process.stdin.setEncoding('utf8');

var choices = {
  // choice => beats
  'paper' : 'rock',
  'rock' : 'scissors',
  'scissors' : 'paper'
};

process.stdout.write("Your Move: ");
process.stdin.on('data', function (input) {
  // calculate computer move
  // source; http://stackoverflow.com/questions/2532218/pick-random-property-from-a-javascript-object
  var computer_move = Object.keys(choices)[ Object.keys(choices).length * Math.random() << 0];

  // get player move, must be valid
  var human_move = input.trim();

  if ( typeof choices[human_move] !== 'undefined' ) {
    if ( human_move == computer_move ) {
      // play until someome wins
      console.log("Tie, Replay!\n");
    } else if ( choices[human_move] == computer_move ) {
      console.log("You Win!\n");
      process.exit();
    } else if ( choices[computer_move] == human_move ) {
      console.log("Computer Wins!\n");
      process.exit();
    }
  }

  process.stdout.write("Your Move: ");
});

